File: android/src/main/java/org/maplibre/reactnative/components/annotation/MarkerViewManager.java
package org.maplibre.reactnative.components.annotation;

import androidx.annotation.NonNull;

import org.maplibre.android.maps.MapView;
import org.maplibre.android.maps.MapLibreMap;

import java.util.ArrayList;
import java.util.List;

import java.lang.reflect.InvocationTargetException;

/**
 * Subclass of MarkerViewManager implementing removeViews and restoreViews
 */
public class MarkerViewManager extends org.maplibre.android.plugins.markerview.MarkerViewManager {
    private final List<MarkerView> markers = new ArrayList<>();
    private MapView mapView;
    private java.lang.reflect.Method markerUpdate;

    public MarkerViewManager(MapView mapView, MapLibreMap map) {
        super(mapView, map);
        this.mapView = mapView;
        // this.map = map;

        try {     
            markerUpdate = MarkerView.class.getSuperclass().getDeclaredMethod("update");
            markerUpdate.setAccessible(true);
        } 
        catch (NoSuchMethodException e) {System.out.println(e.toString());}
    }

    public void addMarker(@NonNull MarkerView markerView) {
        super.addMarker(markerView);
        markers.add(markerView);
    }

    public void removeMarker(@NonNull MarkerView markerView) {
        super.removeMarker(markerView);
        markers.remove(markerView);
    }

    public void removeViews() {
        for (MarkerView marker: markers) {
            mapView.removeView(marker.getView());
        }
    }

    public void restoreViews() {
        for (MarkerView marker: markers) {
            mapView.addView(marker.getView());
        }
    }

    public void updateMarkers(){

        try {
            if (markerUpdate != null) {          
                for( int i = 0; i < markers.size(); i++ ){
                    markerUpdate.invoke(markers.get(i));
                }
            }
        } 
        catch (IllegalArgumentException e) { System.out.println(e.toString()); }
        catch (IllegalAccessException e) { System.out.println(e.toString()); }
        catch (InvocationTargetException e) { System.out.println(e.toString()); }
    }

}

File: package/android/src/main/java/org/maplibre/reactnative/components/annotations/markerview/MarkerViewManager.kt
package org.maplibre.reactnative.components.annotations.markerview

import android.graphics.PointF
import android.graphics.RectF
import android.view.View
import org.maplibre.android.geometry.LatLng
import org.maplibre.android.maps.MapLibreMap
import org.maplibre.android.maps.MapView

class MarkerViewManager(
    private val mapView: MapView,
    private val map: MapLibreMap,
) {
    data class MarkerInfo(
        val view: MLRNMarkerViewContent,
        val markerView: MLRNMarkerView,
        var latLng: LatLng,
        var anchorX: Float = 0f,
        var anchorY: Float = 0f,
        var offsetX: Float = 0f,
        var offsetY: Float = 0f,
    )

    private val markers = mutableListOf<MarkerInfo>()
    private var isDestroyed = false

    fun addMarker(
        view: MLRNMarkerViewContent,
        markerView: MLRNMarkerView,
        latLng: LatLng,
        anchorX: Float = 0f,
        anchorY: Float = 0f,
        offsetX: Float = 0f,
        offsetY: Float = 0f,
    ): MarkerInfo {
        val markerInfo = MarkerInfo(view, markerView, latLng, anchorX, anchorY, offsetX, offsetY)
        markers.add(markerInfo)

        if (view.parent == null) {
            mapView.clipChildren = false
            mapView.clipToPadding = false
            mapView.clipToOutline = false
            mapView.addView(view)
        }

        updateMarkerPosition(markerInfo)

        return markerInfo
    }

    fun removeMarker(markerInfo: MarkerInfo) {
        markers.remove(markerInfo)
        mapView.removeView(markerInfo.view)
    }

    fun removeMarkerByView(view: View) {
        markers.find { it.view == view }?.let { removeMarker(it) }
    }

    fun updateMarkers() {
        if (isDestroyed) return

        for (marker in markers) {
            updateMarkerPosition(marker)
        }
    }

    private fun updateMarkerPosition(marker: MarkerInfo) {
        val view = marker.view
        val screenPos: PointF = map.projection.toScreenLocation(marker.latLng)
        val (viewWidth, viewHeight) = view.getContentSize()
        val anchorOffsetX = viewWidth * marker.anchorX
        val anchorOffsetY = viewHeight * marker.anchorY
        view.x = screenPos.x - anchorOffsetX + marker.offsetX
        view.y = screenPos.y - anchorOffsetY + marker.offsetY
    }

    fun findMarkerByView(view: View): MarkerInfo? = markers.find { it.view == view }

    fun isPointInsideMarker(screenPoint: PointF): Boolean = findMarkerAtPoint(screenPoint) != null

    fun findMarkerAtPoint(screenPoint: PointF): MarkerInfo? {
        for (marker in markers) {
            val v = marker.view
            if (v.visibility != View.VISIBLE) continue
            val (w, h) = v.getContentSize()
            val rect = RectF(v.x, v.y, v.x + w, v.y + h)
            if (rect.contains(screenPoint.x, screenPoint.y)) return marker
        }
        return null
    }

    fun updateMarkerCoordinate(
        markerInfo: MarkerInfo,
        latLng: LatLng,
    ) {
        markerInfo.latLng = latLng
        updateMarkerPosition(markerInfo)
    }

    fun updateMarkerAnchor(
        markerInfo: MarkerInfo,
        anchorX: Float,
        anchorY: Float,
    ) {
        markerInfo.anchorX = anchorX
        markerInfo.anchorY = anchorY
        updateMarkerPosition(markerInfo)
    }

    fun updateMarkerOffset(
        markerInfo: MarkerInfo,
        offsetX: Float,
        offsetY: Float,
    ) {
        markerInfo.offsetX = offsetX
        markerInfo.offsetY = offsetY
        updateMarkerPosition(markerInfo)
    }

    fun removeViews() {
        for (marker in markers) {
            mapView.removeView(marker.view)
        }
    }

    fun restoreViews() {
        for (marker in markers) {
            if (marker.view.parent == null) {
                mapView.addView(marker.view)
            }
        }
        updateMarkers()
    }

    fun onDestroy() {
        isDestroyed = true
        for (marker in markers) {
            mapView.removeView(marker.view)
        }
        markers.clear()
    }
}
